Pi+/Coxeter/Parametrized/FinLehmerEquiv.agda

```diff
@@ -0,0 +1,172 @@
+Coprod-≃-l e = equiv f g f-g g-f
+    where
+    f : _
+    f (inl x) = inl x
+    f (inr x) = inr (–> e x)
+    g : _
+    g (inl x) = inl x
+    g (inr x) = inr (<– e x)
+    f-g : _
+    f-g (inl x) = idp
+    f-g (inr x) = ap inr (<–-inv-r e x)
+    g-f : _
+    g-f (inl x) = idp
+    g-f (inr x) = ap inr (<–-inv-l e x)
+
+Coprod-≃-l-left : {A B C : Type₀} -> (p : A ≃ B) -> (c : C) -> (–> (Coprod-≃-l p) (inl c)) == (inl c)
+Coprod-≃-l-left p c = idp
+
+Coprod-≃-l-right : {A B C : Type₀} -> (p : A ≃ B) -> (a : A) -> (–> (Coprod-≃-l {A} {B} {C} p)) (inr a) == inr (–> p a)
+Coprod-≃-l-right p a = idp
+
+≃-equiv : {A B : Type₀} -> (A ≃ B) -> ((A ≃ A) ≃ (B ≃ B))
+≃-equiv p = TODO
+
+Except : {A : Type₀} -> A -> Type₀
+Except {A} a = Σ A λ b -> ¬ (a == b)
+
+Except≃ : {A : Type₀} -> (has-dec-eq A) -> (a : A) -> (Except a) ⊔ Unit ≃ A
+Except≃ deq a = equiv f g f-g g-f
+    where 
+    f : _
+    f (inl (x , p)) = x
+    f (inr tt) = a
+    g : _
+    g b with deq a b
+    ... | inl _ = inr tt
+    ... | inr x = inl (b , x)
+    f-g : _
+    f-g b with deq a b
+    ... | inl p = p
+    ... | inr x = idp
+    g-f : _
+    g-f (inl (b , p)) with deq a b
+    ... | inl idp = ⊥-elim {_} {λ x → _ == inl (b , p)} (p idp)
+    ... | inr x = ap inl (pair= idp {!  !})
+    g-f (inr tt) with deq a a
+    ... | inl x = idp
+    ... | inr q with (q idp) -- Agda complains when using ⊥-elim...
+    ... | ()
+
+
+bottom-lemma : {false : ⊥} -> {A : Type₀} -> {a : A} -> (⊥-elim false == a)
+bottom-lemma {()}
+
+lemma-f : {A : Type₀} -> Σ ((Unit ⊔ A) ≃ (Unit ⊔ A)) (λ e -> –> e (inl tt) == (inl tt)) -> (A ≃ A)
+lemma-f (e , p) = equiv f g f-g g-f
+    where
+
+    lemma← : <– e (inl tt) == (inl tt)
+    lemma← = ! (! (<–-inv-l e _) ∙ (ap (<– e) p))
+    
+    f : _
+    f a with inspect (–> e (inr a))
+    ... | inl x with== bp = 
+        let t = (! (<–-inv-l e _)) ∙ (ap (<– e) bp)
+        in  ⊥-elim (inl≠inr _ _ (! (t ∙ lemma←)))
+    ... | inr b with== bp = b
+    g : _
+    g a with inspect (<– e (inr a))
+    ... | inl tt with== bp = 
+        let t = (! (<–-inv-r e _)) ∙ (ap (–> e) bp)
+        in  ⊥-elim (inl≠inr _ _ (! (t ∙ p)))
+    ... | inr b with== bp = b
+    f-g : _
+    f-g a with inspect (<– e (inr a)) 
+    ... | inl tt with== bp with (inl≠inr _ _ (! ((! (<–-inv-r e _)) ∙ (ap (–> e) bp) ∙ p)))
+    f-g a | inl tt with== bp | ()
+    f-g a | inr b with== bp with inspect (–> e (inr b))
+    ... | inl tt with== ap' = bottom-lemma
+    ... | inr a' with== ap' = ! (–> (inr=inr-equiv a a') (! ((ap (–> e) (! bp)) ∙ <–-inv-r e (inr a)) ∙ ap'))
+
+    g-f : _
+    g-f a with inspect (–> e (inr a)) 
+    ... | inl tt with== bp with (inl≠inr _ _ (! ((! (<–-inv-l e _)) ∙ (ap (<– e) bp) ∙ lemma←)))
+    g-f a | inl tt with== bp | ()
+    g-f a | inr b with== bp with inspect (<– e (inr b))
+    ... | inl tt with== ap' = bottom-lemma
+    ... | inr a' with== ap' = ! (–> (inr=inr-equiv a a') (! (ap (<– e) (! bp) ∙ (<–-inv-l e (inr a))) ∙ ap'))
```

Comment:
I'd prefer to do these by proving eta-rule like lemmas instead of using inspect.